src/v/storage/segment_deduplication_utils.cc

```diff
@@ -248,4 +251,21 @@ ss::future<index_state> deduplicate_segment(
     co_return new_idx;
 }
 
+ss::future<> linear_scan_of_segment_for_map(
+  const compaction_config& compact_cfg,
+  ss::lw_shared_ptr<segment> seg,
+  key_offset_map& map,
+  probe& pb,
+  model::offset& last_indexed_offset) {
+    co_await map.reset();
+    auto read_holder = co_await seg->read_lock();
+    auto start_offset_inclusive = model::next_offset(last_indexed_offset);
+    auto rdr = internal::create_segment_full_reader(
```

Comment:
Recreating this full segment reader for each round of chunked compaction is a bummer.

 Not sure if we have any abstractions to get around this- [log_reader::reset_config()](https://github.com/redpanda-data/redpanda/blob/f31bbb175be330985a90249c5e590dffdd14368b/src/v/storage/log_reader.h#L163-L176) gave me some hope that the segment's lease/lock could be reused, but it doesn't seem to allow us to reset with a `start_offset` lower than the currently read batch's offset.

For context, we have to do this because in the `chunked_compaction_reducer`, once we fail to index an offset for a record in a batch, we break out of the loop and will have to re-read that batch in the next round using that offset as the start, inclusively.